server/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -632,6 +638,21 @@ private static Object deepCopy(Object value) {
         }
     }
 
+    /**
+     * Executes the given pipeline with for this document unless the pipeline has already been executed
+     * for this document.
+     * @param pipeline Pipeline to execute
+     * @return true if pipeline was executed, false if it could not be executed because it was already executed for this document
+     * @throws Exception On exception in pipeline execution
+     */
+    public boolean executePipeline(Pipeline pipeline) throws Exception {
+        if (this.executedPipelines.add(pipeline) == false) {
```

Comment:
@rjernst I can't really pass the callstack with the document can I? I only have the `execute` method available to pass things (because any called pipeline itself could contain additional pipeline processors).

The reason I made this return `boolean` instead of throwing right away was more of a style thing to make it clear that the exception was triggered by the pipeline processor. But in hindsight this may be a little needlessly complex :) Moving it in here.